torchqdynamics/utils/tensor_types.py

```diff
@@ -24,7 +24,7 @@
 OperatorLike = Union[TensorLike, Qobj]
 
 # type for objects convertible to a `TDOperator` using `time_dependent_to_tensor`
-TDOperatorLike = Union[OperatorLike, Callable[[float], OperatorLike]]
+TDOperatorLike = Union[OperatorLike, Callable[[float], Tensor]]
```

Comment:
> 



> I think we should support the previous typing. It's very handy e.g. for people defining a time-dependent Hamiltonian using QuTiP operators.

I don't think we should allow this. The problem is that, if people pass time-dependent qutip objects, we will have to convert them at every time step. It can significantly slow down the code, especially for device conversion, but even type or dtype conversion is not a good habit.

Hence I would be in favor of not allowing it.

Reply:
> It can significantly slow down the code

I disagree with the _significantly_, I think it's a minor time overhead compared to our gigantic matrix multiplications. I **do agree** that we should raise a warning and have the doc clearly say that it's better to have proper torch tensors for time-dependent Hamiltonians, but for convenience to test things quickly and all, it's quite nice to support qutip objects or numpy arrays. But I'm fine if you really think we should strongly forbid it.

Reply:
No a CPU/GPU transfer of a big matrix is much worse than a single matrix multiplication on GPU. So definitly it can be a significant slowdown. 

However, there are some qutip types that we could allow, such as the `list[list[qobj, callable[[float], complex]]` type. But maybe this should be another PR.

Here is the list of all time-dependent formats supported by qutip: https://qutip.org/docs/latest/guide/dynamics/dynamics-time.html. We can decide on which ones we want to keep or not.

Reply:
Ok my bad then, let's change to forbid non-genuine torch tensors.